torch/distributed/rendezvous.py

```diff
@@ -54,21 +54,26 @@ def rendezvous(url, rank=-1, world_size=-1, **kwargs):
         raise RuntimeError("`world_size` must be an integer. {}".format(world_size))
 
     # Append node-specific arguments.
+    result = urlparse(url)
     if rank != -1 or world_size != -1:
+        query = dict(
+            pair.split("=") for pair in filter(None, result.query.split("&"))
+        )
         assert (
-            "?" not in url
+            "rank" not in query and "world_size" not in query
         ), "The url: {url} has node-specific arguments(rank, world_size) already.".format(
             url=url
         )
-        parts = []
         if rank != -1:
-            parts.append("rank={}".format(rank))
+            query["rank"] = rank
         if world_size != -1:
-            parts.append("world_size={}".format(world_size))
-        if len(parts) > 0:
-            url += "?{parts}".format(parts="&".join(parts))
+            query["world_size"] = world_size
+
+        result = result._replace(
+            query="{}".format("&".join(["{}={}".format(k, v) for k, v in query.items()]))
```

Comment:
The 2 "query"s here have different types. Can we rename the first one to "query_dict" to avoid confusion?